proposals/consolidate-merged-lifecycle-hook-schema.md

```diff
@@ -0,0 +1,126 @@
+
+## Proposal
+
+### A. Update the devcontainer.json schema
+
+#### A1. Change the lifecycle script interface
+
+Add a value (`LifecycleCommandParallel[]`) to the unioned definition of each lifecycle hook (except 'initializeCommand').
+
+> NOTE: _ This was chosen as every existing lifecycle command (and merged version of each command) can be represented within that type._
+
+```typescript
+
+export type LifecycleCommand = string | string[]
+export type LifecycleCommandParallel = { [key: string]: LifecycleCommand; $origin?: string };
+
+interface DevContainerConfig {
+    // ...other devcontainer.json properties...
+	onCreateCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	updateContentCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postCreateCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postStartCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postAttachCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+}
+```
+
+An optional parameter `$origin` is added to the `LifecycleCommandParallel` that supporting tools can use to indicate the source of the command.  For example, this is useful for outputting in the creation log which Feature provided a certain lifecycle hook.  The `$` notation is used to indicate this property is additional tooling metadata that should not be present in a user `devcontainer.json`.
+
+As an example, the following `devcontainer.json` snippet would be valid:
+
+```json
+{
+	"onCreateCommand": [
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "featureA"
+		},
+		{
+			"commandA": "echo 'Hello World!'",
+			"commandB": ["echo", "'Hello World!'"],
+			"$origin": "devcontainer.json"
+		}
+	]
+}
+```
+
+### B. Generating a merged configuration
+
+The [reference implementation](https://github.com/devcontainers/cli) includes a 'read-configuration' command that can be used to resolve a fully merged configuration from a given `devcontainer.json`.
+
+Implementing tools should follow the [documented merging logic](https://containers.dev/implementors/spec/#merge-logic) and output a `devcontainer.json` that represents the resolved configuration.
+
+There are a couple special cases to the merging logic that should be noted:
+
+#### B1. The `customizations` property
+
+The merging of the `customizations` property is left to the implementing tool, therefore there is no defined merging pattern. 
+
+A property `$customizations` should be added to the merged configuration as an array of customizations within each top-level customization object.  Tooling that reads the mergedConfiguration can process this property as needed.
```

Comment:
Why not use a similar approach to the lifecycle commands and support object and array of object values in the `customizations` map's values including `$origin` properties?

Reply:
Thinking about it more I agree that makes sense. Updatesd